lute/io/src/io.cpp

```diff
@@ -0,0 +1,141 @@
+#include "lute/io.h"
+#include "Luau/Variant.h"
+#include "lute/runtime.h"
+
+#include <uv.h>
+#include <memory>
+
+#include "lua.h"
+#include "lualib.h"
+
+
+namespace io
+{
+
+struct IOHandle
+{
+    Luau::Variant<uv_pipe_t, uv_tty_t> streamVariant;
+    uv_stream_t* stream = nullptr;
+    uv_loop_t* loop = nullptr;
+    ResumeToken resumeToken;
+    lua_State* L = nullptr;
+    std::shared_ptr<IOHandle> self;
+    uv_buf_t *buf;
+
+    ~IOHandle() {
+        if (buf) {
+            free(buf->base);
+            buf = nullptr;
+        }
+    }
+
+    void closeHandles()
+    {
+        auto closeCb = [](uv_handle_t* handle)
+        {
+            IOHandle* ioh = static_cast<IOHandle*>(handle->data);
+            ioh->self.reset();
+        };
+
+        uv_read_stop((uv_stream_t*)stream);
+        uv_close((uv_handle_t*)stream, closeCb);
+    }
+};
+
+static void allocBuffer(uv_handle_t* handle, size_t suggestedSize, uv_buf_t* buf)
+{
+    buf->base = (char*)malloc(suggestedSize);
+    buf->len = buf->base ? suggestedSize : 0;
+    if (!buf->base)
+    {
+        fprintf(stderr, "Process pipe buffer allocation failed!\n");
+    }
+    static_cast<IOHandle*>(handle->data)->buf = buf;
+}
+
+// IOHandle is closed immediately after one read since we don't need a long running stream for this API.
+static void onTtyRead(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf)
+{
+    IOHandle* handle = static_cast<IOHandle*>(stream->data);
+
+    if (nread > 0)
+    {
+        handle->resumeToken->complete(
+            [data = std::string(buf->base, nread)](lua_State* L) -> int
+            {
+                lua_pushlstring(L, data.c_str(), data.size());
+                return 1;
+            }
+        );
+    }
+    else if (nread < 0)
+    {
+        handle->resumeToken->fail("Error reading from stdin");
+    }
+    else if (nread == UV_EOF)
+    {
+        handle->resumeToken->fail("EOF reached");
+    }
+
+    handle->closeHandles();
+}
+
+int input(lua_State* L)
+{
+    auto handle = std::make_shared<IOHandle>();
+    handle->loop = uv_default_loop();
+    handle->resumeToken = getResumeToken(L);
+    handle->L = L;
+    handle->self = handle;
+
+    uv_handle_type ht = uv_guess_handle(fileno(stdin));
+    if (ht == UV_TTY)
+    {
+        uv_tty_t& tty = handle->streamVariant.emplace<uv_tty_t>();
+        handle->stream = (uv_stream_t*)&tty;
+        uv_tty_init(handle->loop, (uv_tty_t*)&tty, fileno(stdin), 0);
```

Comment:
What happens if this call fails?

Reply:
I think we'll want to signal an error here - something like failed to init tty.

Reply:
It returns a status code like other libuv APIs, so we can check if it's negative and report the stringified error if so.

Reply:
We should use `luaL_error` here too, something like
```suggestion
        if (int status = uv_tty_init(handle->loop, (uv_tty_t*)&tty, fileno(stdin), 0); status < 0)
            luaL_error(uv_strerror(status));
```